ps_crossselling.php

```
@@ -148,200 +216,194 @@ public function renderForm()
+            if (!empty($products)) {
+                foreach ($products as $product) {
+                    $productIds[] = (int) $product['id_product'];
+                }
+            }
+        }
 
-    private function getProductIds($hookName, array $configuration)
-    {
-        if ('displayShoppingCart' === $hookName || 'displayShoppingCartFooter' === $hookName) {
-            $productIds = array_map(function ($elem) {
-                return $elem['id_product'];
-            }, $configuration['cart']->getProducts());
-        } else {
-            $productIds = [$configuration['product']['id_product']];
+        if (!empty($configuration['product']['id_product']) && !in_array($configuration['product']['id_product'], $productIds)) {
+            $productIds[] = (int) $configuration['product']['id_product'];
         }
 
-        return array_unique($productIds);
+        return $productIds;
     }
 
-    public function getWidgetVariables($hookName, array $configuration)
+    /**
+     * @param int[] $productIds
+     *
+     * @return int[]
+     */
+    protected function getOrderProducts(array $productIds = [])
     {
-        $productIds = $this->getProductIds($hookName, $configuration);
-        if (!empty($productIds)) {
-            $products = $this->getOrderProducts($productIds);
+        $orderedProductIds = [];
+        $db = Db::getInstance((bool) _PS_USE_SQL_SLAVE_);
 
-            if (!empty($products)) {
-                return [
-                    'products' => $products,
-                ];
+        try {
+            /** @var FilesystemCache $cache */
+            $cache = $this->get('ps_crossselling.cache.products');
+            $cacheKey = $this->getCacheId() . '|' . implode('|', $productIds);
+
+            if ($cache->has($cacheKey)) {
+                return $cache->get($cacheKey);
             }
+        } catch (Exception $exception) {
+            $cache = null;
+            $cacheKey = null;
```

I'm not sure it's a good idea to catch all exceptions and continue silently. The cache might be failing and you won't be aware. As a general rule, only catch the exact exception you know you can get back from.

---

My goal was to avoid cache issue with container with some agressive cache.
https://github.com/PrestaShopCorp/ps_mbo/pull/96
So I will catch only ServiceNotFoundException